projects/telemetry/src/gps.c

```diff
@@ -0,0 +1,113 @@
+#include "gps.h"
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include "delay.h"
+#include "gpio.h"
+#include "interrupt.h"
+#include "nmea.h"
+#include "status.h"
+#include "uart.h"
+
+// Two structs to store data and settings. Since the GPS should only be initialized once
+static GpsSettings *s_settings = NULL;
+static GpsStorage *s_storage = NULL;
+
+// This method will be called every time the GPS sends data.
+static void prv_gps_callback(const uint8_t *rx_arr, size_t len, void *context) {
+  NmeaMessageId messageId = NMEA_MESSAGE_ID_UNKNOWN;
+  nmea_sentence_type((char *)rx_arr, &messageId);
+  if (messageId == NMEA_MESSAGE_ID_GGA) {         // GGA message
+    strncpy((char *)s_storage->gga_data, (char *)rx_arr, GPS_MAX_NMEA_LENGTH);
+  } else if (messageId == NMEA_MESSAGE_ID_VTG) {  // VTG message
+    strncpy((char *)s_storage->vtg_data, (char *)rx_arr, GPS_MAX_NMEA_LENGTH);
+  }
+}
+
+// The GPS power line will be connected to a 3V pin. This method will
+// control if that pin is providing power or not.
+static void prv_gps_set_power_state(bool powered) {
+  if (powered) {
+    gpio_set_state(s_settings->pin_power, GPIO_STATE_HIGH);
+  } else {
+    gpio_set_state(s_settings->pin_power, GPIO_STATE_LOW);
+  }
+}
+
+// Stage 1 of initialization: pull high on ON/OFF line
+static void prv_gps_init_stage_1() {
+  gpio_set_state(s_settings->pin_on_off, GPIO_STATE_HIGH);
+}
+
+// Stage 2: set ON/OFF line to low
+static void prv_gps_init_stage_2() {
+  gpio_set_state(s_settings->pin_on_off, GPIO_STATE_LOW);
+}
+
+// Initialization of this chip is described on page 10 of:
+// https://www.linxtechnologies.com/wp/wp-content/uploads/rxm-gps-f4.pdf
+StatusCode gps_init(GpsSettings *settings, GpsStorage *storage) {
+  if (s_settings != NULL) {
+    return status_msg(STATUS_CODE_RESOURCE_EXHAUSTED, "Cannot reinitialize GPS\n");
+  }
+
+  // Initialize both structs
+  s_settings = settings;
+  s_storage = storage;
+
+  GpioSettings telemetry_settings_gpio_general = {
+    .direction = GPIO_DIR_OUT,  // The pin needs to output.
+    .state = GPIO_STATE_LOW,    // Start in the "off" state.
+    .alt_function = GPIO_ALTFN_1,
+  };
+
+  // Initializes UART
+  StatusCode ret =
```

Comment:
This return variable is never used? Also might want to either change the signature of the init method, or return useful status codes instead of just STATUS_CODE_OK at the bottom